# Single Cell Gene Expression Analysis Workflows

This folder contains the analysis workflows of a Flex Gene Expression demo dataset taken from 10x Genomics.

## Dataset overview

The **Kidney and Lung DTCs Matched PBMC MultiProÂ® Human Discovery Panel** dataset was generated by using Gene Expression Flex. 
Please follow [this link](https://www.10xgenomics.com/datasets/160k_DTC_Matched_PBMC_MultiPro_Human_Discovery_Panel) 
to the 10x Genomics Datasets website for more information and download the Cell Ranger **per-sample** outputs.

For this demo, you will need the <ins>filtered</ins> and <ins>raw</ins> feature/cell matrix HDF5 files from:
- PBMC Control 1
  -  `160k_DTC_Matched_PBMC_MultiPro_Human_Discovery_Panel_PBMC_Control1_count_sample_filtered_feature_bc_matrix.h5`
  -  `160k_DTC_Matched_PBMC_MultiPro_Human_Discovery_Panel_PBMC_Control1_count_sample_raw_feature_bc_matrix.h5`
- Lung Cancer PBMCs
  - `160k_DTC_Matched_PBMC_MultiPro_Human_Discovery_Panel_Lung_Cancer_PBMC_count_sample_filtered_feature_bc_matrix.h5`
  - `160k_DTC_Matched_PBMC_MultiPro_Human_Discovery_Panel_Lung_Cancer_PBMC_count_sample_raw_feature_bc_matrix.h5`
- Kidney Cancer PBMCs
  - `160k_DTC_Matched_PBMC_MultiPro_Human_Discovery_Panel_Kidney_Cancer_PBMC_count_sample_filtered_feature_bc_matrix.h5`
  - `160k_DTC_Matched_PBMC_MultiPro_Human_Discovery_Panel_Kidney_Cancer_PBMC_count_sample_raw_feature_bc_matrix.h5`

## File types

In each folder, the `no-output.ipynb` files are Jupyter Notebook files with all output text and images removed. 
This is to avoid the `ipynb` files from getting too big.

The `.md` files contains the identical workflow but in the Markdown syntax. 
This allow you to view the workflow including outputs on GitHub.

## Using the workflows

First, start the single-sample analysis with the HDF5 files from one of the three samples, for example Control 1 (see `160k_Human_Control1_PBMC`). 
At the end of the said analysis, the final `SingleCellExperiment` containing the filtered cells and any added metadata is stored as HDF5-based object.

Then, repeat the single-sample analysis with the remaining two PBMC samples.

Finally, use the integration workflow found in the `160k_Human_Integrated` folder 
to combine the cells from all three samples into a common space for consistent analyses 
and to perform differential expression comparisons.

## References

The [Orchestrating Single-Cell Analysis with Bioconductor](https://bioconductor.org/books/3.20/OSCA/) 
provides an invaluable resource to the creation and development of the workflows presented here.
